File: Lab-3/source/bai5.java
package source;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.Mapper;
import org.apache.hadoop.mapreduce.Reducer;

import org.apache.hadoop.mapreduce.lib.input.MultipleInputs;
import org.apache.hadoop.mapreduce.lib.input.TextInputFormat;

import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;

import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.FSDataInputStream;

public class bai5 {

    public static class OccupationRatingMapper
            extends Mapper<Object, Text, Text, Text> {

        // UserID → OccupationID
        private final Map<String, String> userOccupationMap = new HashMap<>();

        // OccupationID → OccupationName
        private final Map<String, String> occupationNameMap = new HashMap<>();

        private final Text outKey = new Text();
        private final Text outValue = new Text();

        @Override
        protected void setup(Context context)
                throws IOException, InterruptedException {

            Configuration conf = context.getConfiguration();

            String usersPath = conf.get("users.path");
            String occupationsPath = conf.get("occupations.path");

            FileSystem fs = FileSystem.get(conf);

            // =========================
            // READ users.txt
            // =========================
            try (
                FSDataInputStream in = fs.open(new Path(usersPath));
                BufferedReader br = new BufferedReader(
                        new InputStreamReader(in))
            ) {

                String line;

                while ((line = br.readLine()) != null) {

                    line = line.trim();

                    if (line.isEmpty()) {
                        continue;
                    }

                    // bỏ header nếu có
                    if (line.toLowerCase().startsWith("userid")) {
                        continue;
                    }

                    // UserID,Gender,Age,OccupationID,Zip
                    String[] parts = line.split(",");

                    if (parts.length >= 4) {

                        String userId = parts[0].trim();
                        String occupationId = parts[3].trim();

                        userOccupationMap.put(userId, occupationId);
                    }
                }
            }

            // =========================
            // READ occupation.txt
            // =========================
            try (
                FSDataInputStream in = fs.open(new Path(occupationsPath));
                BufferedReader br = new BufferedReader(
                        new InputStreamReader(in))
            ) {

                String line;

                while ((line = br.readLine()) != null) {

                    line = line.trim();

                    if (line.isEmpty()) {
                        continue;
                    }

                    // bỏ header nếu có
                    if (line.toLowerCase().startsWith("occupationid")
                            || line.toLowerCase().startsWith("id")) {
                        continue;
                    }

                    // OccupationID,OccupationName
                    String[] parts = line.split(",", 2);

                    if (parts.length >= 2) {

                        String occId = parts[0].trim();
                        String occName = parts[1].trim();

                        occupationNameMap.put(occId, occName);
                    }
                }
            }

            System.out.println("Loaded Users: " + userOccupationMap.size());
            System.out.println("Loaded Occupations: " + occupationNameMap.size());
        }

        @Override
        public void map(Object key, Text value, Context context)
                throws IOException, InterruptedException {

            String line = value.toString().trim();

            if (line.isEmpty()) {
                return;
            }

            // bỏ header nếu có
            if (line.toLowerCase().startsWith("userid")) {
                return;
            }

            // UserID,MovieID,Rating,Timestamp
            String[] parts = line.split(",");

            if (parts.length < 3) {
                return;
            }

            try {

                String userId = parts[0].trim();

                double rating =
                        Double.parseDouble(parts[2].trim());

                // lookup occupation id
                String occupationId =
                        userOccupationMap.get(userId);

                if (occupationId == null) {
                    return;
                }

                // lookup occupation name
                String occupationName =
                        occupationNameMap.get(occupationId);

                if (occupationName == null) {
                    occupationName = "Unknown";
                }

                // emit
                outKey.set(occupationName);
                outValue.set(rating + ",1");

                context.write(outKey, outValue);

            } catch (Exception e) {
                // pass
            }
        }
    }

    public static class OccupationRatingReducer
            extends Reducer<Text, Text, Text, Text> {

        private final Text result = new Text();

        @Override
        public void reduce(Text key,
                           Iterable<Text> values,
                           Context context)
                throws IOException, InterruptedException {

            double totalRating = 0.0;
            int totalCount = 0;

            for (Text val : values) {

                String[] parts =
                        val.toString().split(",");

                if (parts.length < 2) {
                    continue;
                }

                try {

                    double rating =
                            Double.parseDouble(parts[0]);

                    int count =
                            Integer.parseInt(parts[1]);

                    totalRating += rating;
                    totalCount += count;

                } catch (Exception e) {
                    // pass
                }
            }

            if (totalCount == 0) {
                return;
            }

            double avgRating =
                    totalRating / totalCount;

            result.set(
                    String.format("Average Rating: %.2f\tTotal Ratings: %d",
                            avgRating,
                            totalCount)
            );

            context.write(key, result);
        }
    }

    public static void main(String[] args)
            throws Exception {

        if (args.length < 5) {

            System.err.println(
                    "Usage: bai5 <ratings1> <ratings2> <users> <occupations> <output>"
            );

            System.exit(1);
        }

        Configuration conf = new Configuration();

        conf.set("users.path", args[2]);
        conf.set("occupations.path", args[3]);

        Job job = Job.getInstance(
                conf,
                "bai5 - Rating Analysis by Occupation"
        );

        job.setJarByClass(bai5.class);

        job.setMapperClass(OccupationRatingMapper.class);
        job.setReducerClass(OccupationRatingReducer.class);

        job.setMapOutputKeyClass(Text.class);
        job.setMapOutputValueClass(Text.class);

        job.setOutputKeyClass(Text.class);
        job.setOutputValueClass(Text.class);

        MultipleInputs.addInputPath(
                job,
                new Path(args[0]),
                TextInputFormat.class,
                OccupationRatingMapper.class
        );

        MultipleInputs.addInputPath(
                job,
                new Path(args[1]),
                TextInputFormat.class,
                OccupationRatingMapper.class
        );

        Path outputPath = new Path(args[4]);

        FileSystem fs = FileSystem.get(conf);

        if (fs.exists(outputPath)) {

            fs.delete(outputPath, true);

            System.out.println(
                    "[Bai5] Deleted old output directory: "
                            + outputPath
            );
        }

        FileOutputFormat.setOutputPath(job, outputPath);

        System.exit(job.waitForCompletion(true) ? 0 : 1);
    }
}

File: Lab-3/source/bai5.py
#!/usr/bin/env python3
"""Bài 5 - Phân tích rating trung bình và tổng lượt đánh giá theo occupation bằng PySpark RDD.

Usage:
    spark-submit source/bai5.py <ratings_1> <ratings_2> <users> <occupation> <output>
"""

from __future__ import annotations

import sys
from pyspark import SparkConf, SparkContext

from common import parse_rating, parse_user, parse_occupation, add_rating_stats, average, save_text_rdd


def build_result(sc: SparkContext, ratings1_path: str, ratings2_path: str, users_path: str, occupations_path: str):
    occupation_names = (
        sc.textFile(occupations_path)
        .map(parse_occupation)
        .filter(lambda row: row is not None)
    )  # occupation_id -> occupation_name

    # First build user_id -> occupation_name using only RDD joins.
    user_id_by_occupation_id = (
        sc.textFile(users_path)
        .map(parse_user)
        .filter(lambda row: row is not None)
        .map(lambda row: (row[3], row[0]))
    )

    user_occupation_name = (
        user_id_by_occupation_id
        .join(occupation_names)  # occupation_id -> (user_id, occupation_name)
        .map(lambda item: (item[1][0], item[1][1]))
    )

    ratings_by_user = (
        sc.textFile(ratings1_path)
        .union(sc.textFile(ratings2_path))
        .map(parse_rating)
        .filter(lambda row: row is not None)
        .map(lambda row: (row[0], row[2]))
    )

    return (
        ratings_by_user
        .join(user_occupation_name)  # user_id -> (rating, occupation_name)
        .map(lambda item: (item[1][1], (item[1][0], 1)))
        .reduceByKey(add_rating_stats)
        .mapValues(lambda stats: (average(stats), stats[1]))
        .sortBy(lambda item: item[0])
        .map(lambda item: f"{item[0]}\tAverage Rating: {item[1][0]:.2f}\tTotal Ratings: {item[1][1]}")
    )


def main(argv: list[str]) -> int:
    if len(argv) != 6:
        print("Usage: bai5.py <ratings_1> <ratings_2> <users> <occupation> <output>", file=sys.stderr)
        return 1

    ratings1_path, ratings2_path, users_path, occupations_path, output_path = argv[1:6]
    conf = SparkConf().setAppName("Lab3-Bai5-Occupation-Rating-RDD")
    sc = SparkContext(conf=conf)
    try:
        result = build_result(sc, ratings1_path, ratings2_path, users_path, occupations_path)
        save_text_rdd(result, output_path)
    finally:
        sc.stop()
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))
